### ISSUE TITLE
Is it possible to convert a cid to an int inferior to 2**256-1?

### ISSUE THREAD

**Issue by n1c01a5** [multiformats/js-cid-tool#12]
I ask this question because the ERC721 works with an integer, see https://github.com/OpenZeppelin/openzeppelin-contracts/blob/406c83649bd6169fc1b578e08506d78f0873b276/contracts/token/ERC721/ERC721.sol#L91 .

So how I can convert an IPFS CID like QmRv33GhnvB5RzRizarXhGxayTBWu6kFWuf6VZhWYGfhog to an integer without an overflow?

**Comment by lidel** [comment]
@n1c01a5 `QmRv33GhnvB5RzRizarXhGxayTBWu6kFWuf6VZhWYGfhog` is a CIDv0 with sha2-256 Multihash.
You can inspect what each byte means at: https://cid.ipfs.io/#QmRv33GhnvB5RzRizarXhGxayTBWu6kFWuf6VZhWYGfhog

The answer to your question depends on if your code needs to be future-proof.
If you are okay with **hardcoding use of sha2-256 in your app**, then you could extract digest  from the CID and assume it is always  256 bit.

You can extract sha2-256 digest without additional prefixes via (base16 will return HEX values):


```console
$ ipfs cid format QmRv33GhnvB5RzRizarXhGxayTBWu6kFWuf6VZhWYGfhog -f "%D" -b base16
35207a00e0b36c403b57b9f84231ff643f6aea2eab78a1dabafda6a9f707d1b3
```

Regarding future-proofing your code, you need to either support arbitrary CIDs, or reject everything longer than 256 bit digest. This is a real concern  because in the future people may use CIDv1 with sha2-512 which has 512 bit digest:

```console
$ ipfs cid format bafkrgqhhyivzstcz3hhswshfjgy6ertgmnqeleynhwt4dlfsthi4hn7zgh4uvlsb5xncykzapi3ocd4lzogukir6ksdy6wzrnz6ohnv4aglcs -f "%D" -b base16
e7c22b994c59d9cf2b48e549b1e24666636045930d3da7c1acb299d1c3b7f931f94aae41edda2c2b207a36e10f8bcb8d45223e54878f5b316e7ce3b6bc019629
```

TLDR: if you want your code to be future-proof,  support CIDv1 ([specs](https://github.com/multiformats/cid)). IF not, make sure you reject CIDs with longer hashes to avoid hard to debug errors when people start using them.

ps. In the future, ask support questions at forums at https://discuss.ipfs.io/c/help/13 where more people can see it :-)
